Speedups for precommit hooks

eparis opened the pull request:
This (along with the separate munger change) greatly reduces the pre-commit hook time. From like 30+ seconds down to 8.5 seconds to build everything, 2.5 seconds  to get swagger (that stinks) and another 2 seconds to do everything else in the hook.

Notice this does make it easier for the gendeepcopy and genconversion to be wrong when run by hand. This is the exact same as gen\* can be wrong when run by hand if you didn't build new copies.  The git pre-commit hook builds everything, so it should always be right (as would travis et al once you open a PR)

eparis committed:
Use cp instead of rsync in hack/verify-gendocs.sh

Instead of calling rsync over and over and over and over and over and
over and over and over and over (and probably over) use one `cp`

Before:
real	0m5.247s
user	0m2.294s
sys	0m1.300s

After:
real	0m2.260s
user	0m2.230s
sys	0m0.936s

eparis committed:
Use hack/verify-boilerplate.sh in pre-commit hook

Using the hack/ version is probably a little slower, but it still only
takes about .2 seconds. So probably worth the reduction in code.

eparis committed:
Split hack/{verify,update}-* files so we don't always go build

Right now some of the hack/* tools use `go run` and build almost every
time. There are some which expect you to have already run `go install`.
And in all cases the pre-commit hook, which runs a full build wouldn't
want to do either, since it just built!

This creates a new hack/after-build/ directory and has the scripts which
REQUIRE that the binary already be built. It doesn't test and complain.
It just fails miserably. Users should not be in this directory. Users
should just use hack/verify-* which will just do the build and then call
the "after-build" version. The pre-commit hook or anything which KNOWS
the binaries have been built can use the fast version.

eparis committed:
Update swagger to start etcd with known values

We were setting etcd values, but we started etcd before we set them.
Thankfully we were just setting them to etcd defaults, so it all worked

eparis committed:
hack/lib/etcd: remove unused local var

It isn't used anywhere, why have it?

mikedanese commented:
@thockin

zmerlynn commented:
So this now even more assumes that your local build is up to date, it looks like? i.e. if you haven't done a `make` in a little while, your precommits might just fail? (which sucks for some workflows, like `git bisect`)

thockin commented on hack/lib/util.sh:
```diff
@@ -171,7 +171,11 @@ kube::util::gen-doc() {
-  find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+  files=($(find "${tmpdir}" -type f))
+  if [[ "${#files[@]}" -ne 0 ]]; then
+    cp -f "${files[@]}" "${dest}"
```
I don't think this does the right thing.  Consider:

```
thockin@freakshow:/tmp$ rm -rf a z

thockin@freakshow:/tmp$ mkdir -p a/b/c/d

thockin@freakshow:/tmp$ touch a/b/c/d/file-in-a-b-c-d

thockin@freakshow:/tmp$ touch a/b/c/file-in-a-b-c

thockin@freakshow:/tmp$ mkdir -p z/b/c/d

thockin@freakshow:/tmp$ find a -type f
a/b/c/d/file-in-a-b-c-d
a/b/c/file-in-a-b-c

thockin@freakshow:/tmp$ find z -type f

thockin@freakshow:/tmp$ f=($(find a -type f))

thockin@freakshow:/tmp$ cp -f "${f[@]}" z

thockin@freakshow:/tmp$ find z -type f
z/file-in-a-b-c
z/file-in-a-b-c-d
```

thockin commented on hooks/pre-commit:
```diff
@@ -41,31 +41,13 @@ fi
-fi
-
-if [[ "${#files_need_boilerplate[@]}" -ne 0 ]]; then
+out=($(hack/verify-boilerplate.sh))
```
Won't this catch files that may have snuck in but are not part of this commit?  Maybe not a huge deal, but we could pass "${files[@]}" to the script and have it use the commandline if present, else find files itself

eparis replied:
It will. I've got some other branch to make that possible, but for now, just doing it for the whole repo is really fast.  I can pull this out if you want...

thockin commented on hack/update-generated-conversions.sh:
```diff
@@ -44,7 +56,7 @@ import (
 // AUTO-GENERATED FUNCTIONS START HERE
 EOF
 
-	GOPATH=$(godep path):$GOPATH go run cmd/genconversion/conversion.go -v ${version} -f - >>  $TMPFILE
```
There's a reason we force preconditions by 'go run' - we have been bitten over and over by stale builds of tools "passing" precommit, only to fail CI (or worse, be committed).

I think the right pattern to follow is something like:

update-generated-conversions.sh:
```
#!/bin/bash
make WHAT=cmd/genconversion
exec $(realpath $(dirname $0))/update-generated-conversions__prebuilt.sh
```

update-generated-conversions__prebuilt.sh
```
#!/bin/bash
# test for existence of the binary/binaries we need
# run them
```

Then have precommit run a `make` followed by all of the __prebuilt scripts. Humans run the regular scripts.

thockin commented:
Eric,

Are you still working on this.  It's driving me batty a little more each day...

eparis commented:
i keep hoping to get back to it every afternoon.  today looks ok.

eparis committed:
Move boilerplate from hooks/ to hack/

Hooks seems like they should be hooks. Not sure hack makes a lot more
sense, but it has more stuff already.

eparis committed:
Move description.sh from hack/ to hooks/

Just to get everything in one place...

eparis commented:
hopefully other comments have been addressed.  PTAL.

thockin commented:
LGTM

eparis commented:
@k8s-bot test this

cjcullen commented:
@eparis Do you want to squash commits before I merge?

eparis commented:
squashed as seemed appropriate...  each commit should be its own thing now.

cjcullen commented:
Thanks!
